SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/Components/SpatialPingComponent.h"
+
+#include "GameFramework/PlayerController.h"
+#include "Net/UnrealNetwork.h"
+#include "TimerManager.h"
+#include "Engine/World.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialPingComponent);
+
+USpatialPingComponent::USpatialPingComponent(const FObjectInitializer& ObjectInitializer)
+	: Super(ObjectInitializer)
+{
+	PrimaryComponentTick.bCanEverTick = false;
+	PrimaryComponentTick.bStartWithTickEnabled = false;
+
+	bReplicates = true;
+}
+
+void USpatialPingComponent::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME_CONDITION(USpatialPingComponent, ReplicatedPingID, COND_OwnerOnly);
+}
+
+void USpatialPingComponent::BeginPlay()
+{
+	Super::BeginPlay();
+	// Attempt to cast the owner of this component to the PlayerController class.
+	// Component will do nothing if cast fails.
+	OwningController = Cast<APlayerController>(GetOwner());
+	if (OwningController == nullptr)
+	{
+		UE_LOG(LogSpatialPingComponent, Warning, TEXT("SpatialPingComponent did not find a valid owning PlayerController and will not function correctly. Ensure this component is only attached to a PlayerController."));
+	}
+
+	if (bStartWithPingEnabled)
+	{
+		SetPingEnabled(true);
+	}
+}
+
+void USpatialPingComponent::EndPlay(const EEndPlayReason::Type EndPlayReason)
+{
+	// Clear up the timer before the component is disabled or destroyed.
+	SetPingEnabled(false);
+	Super::EndPlay(EndPlayReason);
+}
+
+bool USpatialPingComponent::GetIsPingEnabled() const
+{
+	return bIsPingEnabled;
+}
+
+void USpatialPingComponent::SetPingEnabled(bool bSetEnabled)
+{
+	// Only execute if this component is attached to a player controller.
+	if (OwningController != nullptr)
```

Comment:
Nit: invert this if to reduce indentation.
```suggestion
	if (OwningController == nullptr)
	{
		return;
	}
```